### ISSUE TITLE
How can i plot the loss and mAP diagram after training yolov8 ?

### ISSUE THREAD

**Issue by yuritt55** [ultralytics/ultralytics#13859]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

How can i plot the loss and mAP diagram after training yolov8 ?

**Comment by glenn-jocher** [comment]
Hello @yuritt55,

Thank you for your question! To plot the loss and mAP diagrams after training YOLOv8, you can utilize the training logs saved by the model. Here's a step-by-step guide:

1. **Using TensorBoard**:
   - YOLOv8 integrates seamlessly with TensorBoard, which is a powerful tool for visualizing training metrics.
   - After training, you can launch TensorBoard from your terminal:
     ```bash
     tensorboard --logdir runs
     ```
   - Open the provided URL in your browser to view the loss and mAP diagrams.

2. **Using Matplotlib**:
   - If you prefer to use Python for custom plots, you can extract the metrics from the training logs and plot them using Matplotlib. Here's a basic example:
     ```python
     import matplotlib.pyplot as plt
     import yaml

     # Load training results
     with open('runs/train/exp/results.yaml', 'r') as f:
         results = yaml.safe_load(f)

     # Extract metrics
     epochs = range(len(results['metrics']['box']['map50']))
     loss = results['metrics']['loss']
     map50 = results['metrics']['box']['map50']

     # Plot Loss
     plt.figure()
     plt.plot(epochs, loss, label='Loss')
     plt.xlabel('Epochs')
     plt.ylabel('Loss')
     plt.title('Training Loss')
     plt.legend()
     plt.show()

     # Plot mAP
     plt.figure()
     plt.plot(epochs, map50, label='mAP@0.5')
     plt.xlabel('Epochs')
     plt.ylabel('mAP@0.5')
     plt.title('Mean Average Precision')
     plt.legend()
     plt.show()
     ```

These methods should help you visualize the training progress effectively. If you encounter any issues or have further questions, feel free to ask!

**Comment by glenn-jocher** [comment]
Hello @Mary14-design,

Thank you for reaching out! Improving training and validation results can often be achieved by fine-tuning various aspects of your training process. Here are a few tips to help you get started:

1. **Hyperparameter Tuning**: While you mentioned not modifying hyperparameters, adjusting them can significantly impact performance. Consider experimenting with:
   - Learning rate (`lr0`)
   - Batch size (`batch`)
   - Number of epochs (`epochs`)

2. **Data Augmentation**: Enhancing your dataset with augmentation techniques can improve model robustness. You can enable and configure augmentations in your training script or YAML file.

3. **Model Architecture**: If YOLOv8s is not yielding the desired results, you might want to try a different model variant (e.g., YOLOv8m or YOLOv8l) to see if a larger model performs better.

4. **Training Duration**: Sometimes, extending the training duration can help the model converge better. Ensure that you are training for a sufficient number of epochs.

5. **Validation Frequency**: Regularly validate your model during training to monitor its performance and make adjustments as needed.

Here's an example of how you can modify the training command to include some of these suggestions:
```python
from ultralytics import YOLO

# Load a model
model = YOLO('yolov8s.yaml')  # build a new model from YAML
model = YOLO('yolov8s.pt')  # load a pretrained model (recommended for training)

# Train the model with modified hyperparameters
results = model.train(data='coco8.yaml', epochs=100, imgsz=640, lr0=0.01, batch=16)
```

Additionally, please ensure you are using the latest version of the Ultralytics package to benefit from the latest features and bug fixes.

For more detailed guidance, you can refer to our [Training Guide](https://docs.ultralytics.com/modes/train).

Feel free to reach out if you have any further questions or need additional assistance. Happy training! 🚀

**Comment by Mary14-design** [comment]
Thanks a lot for the quick feedback. Any clue why the mAP values stay constant?


![image](https://github.com/ultralytics/ultralytics/assets/54886321/c8b6d355-2165-47f7-92f7-b9dbd713dbc3)

**Comment by glenn-jocher** [comment]
Hello @Mary14-design,

Thank you for your follow-up! The mAP values staying constant can be indicative of a few potential issues. Here are some steps you can take to diagnose and address this:

1. **Learning Rate**: Ensure that your learning rate is not too high or too low. An inappropriate learning rate can cause the model to get stuck in a local minimum or not learn effectively. You might want to try adjusting the `lr0` parameter.

2. **Data Quality**: Verify that your dataset is correctly labeled and that there are no issues with the data itself. Poor quality or inconsistent data can lead to stagnant mAP values.

3. **Overfitting**: If your model is overfitting, it might perform well on the training data but not improve on the validation data. Consider using techniques like data augmentation, dropout, or regularization.

4. **Validation Frequency**: Ensure that you are validating your model frequently enough to capture improvements. You can adjust the validation frequency in your training script.

5. **Model Architecture**: Sometimes, the chosen model architecture might not be suitable for the task or dataset. Experimenting with different YOLOv8 variants (e.g., YOLOv8m or YOLOv8l) might yield better results.

Here's an example of how you can adjust the learning rate and batch size in your training script:
```python
from ultralytics import YOLO

# Load a model
model = YOLO('yolov8s.pt')  # load a pretrained model (recommended for training)

# Train the model with modified hyperparameters
results = model.train(data='coco8.yaml', epochs=100, imgsz=640, lr0=0.01, batch=16)
```

If the issue persists, please provide a minimum reproducible example to help us better understand and diagnose the problem. You can find more details on how to create one [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Lastly, ensure you are using the latest version of the Ultralytics package to benefit from recent updates and fixes.

Feel free to reach out if you have any further questions or need additional assistance. Happy training! 🚀
